src/DataAPI.jl

```diff
@@ -287,4 +287,100 @@ using a `sink` function to materialize the table.
 """
 function allcombinations end
 
+const STYLE_INFO = """
+One of the uses of the metadata style information is decision
+how the metadata should be propagated when `x` is transformed. This interface
+defines `:none` style that indicates that metadata should not be propagated
+under transformations. At least this style must be supported by any type
+defining support for metadata.
+"""
+
+"""
+    metadata(x, key, [default]; full::Bool=false)
+
+Return metadata value associated with object `x` for key `key`.
+If `x` does not support metadata throw `ArgumentError`.
+If `x` supports metadata, but does not have a mapping for `key` throw `KeyError`.
+
+If `full=true` return a tuple of metadata value and metadata style. Metadata
+style is an additional information about the kind of metadata that is stored
+for the `key`.
+
+$STYLE_INFO
+
+If `default` is passed then return it if `x` does not support metadata or
+does not have a mapping for `key`. The style of `default` value is `:none`.
```

Comment:
Shouldn't we also provide  `hasmetadata(x)`, `hasmetadata(x, key)` and `hascolmetadata(x)`, `hascolmetadata(x, col)` and `hascolmetadata(x, col, key)`?

Reply:
We do not need. if `metadatakeys` returns `()` we know there is no metadata. Otherwise to check if metadata for some `key` is present user should use `key in metadata(x)`. Same with column metadata.